### PR TITLE
Made subsurface tests parameterized

### PR DESCRIPTION
**Pull request by wmww**
Tests subsurfaces
* On WL shell surfaces with mouse input
* On XDG shell surfaces with mouse input
* On default shell surfaces with touch input

### CODE REVIEW

File: tests/subsurfaces.cpp [hunk 1]
```diff
@@ -26,16 +26,103 @@
...
+
+struct PointerInputDevice : AbstractInputDevice
+{
+    PointerInputDevice(wlcs::Server& server, wlcs::Client& client):
+        client{client},
+        pointer{server.create_pointer()}
+    {
+    }
+
+    void down_at(int x, int y) override
```
Comment (AlanGriffiths) [inline tests/subsurfaces.cpp]: The name doesn't seem to match the implementation. Which is right?

Comment (wmww) [reply tests/subsurfaces.cpp]: It's an override of an abstract method. 'down_at' doesn't make sense for pointers and 'move_to' doesn't make sense for touch. Maybe there's a name out there that's applicable to both, but I just choose one and went with it.

Comment (AlanGriffiths) [reply tests/subsurfaces.cpp]: place_at()?

Comment (wmww) [reply tests/subsurfaces.cpp]: Maybe `to_screen_position` is a bit better

**Pull request merged**